vignettes/lifecycle.Rmd

```diff
@@ -12,246 +12,341 @@ knitr::opts_chunk$set(
+#' 
+#' In brief,
+#' `df %>% gather("key", "value", x, y, z)` is equivalent to
+#' `df %>% pivot_longer(c(x, y, z), names_to = "key", values_to = "value")`.
+#' See more details in `vignette("pivot")`.
+```
+
+The rest of the documentation can stay the same.
+
+## Arguments
 
-# Force warnings
-options(lifecycle_verbosity = "warning")
+### Deprecating, keeping existing default
 
-# Force errors
-options(lifecycle_verbosity = "error")
+Take this example where we want to deprecate `na.rm` in favour of always making it `TRUE.`
+
+```{r}
+add_two <- function(x, y, na.rm = TRUE) {
+  sum(x, y, na.rm = na.rm)
+}
 ```
 
-Forcing warnings can be useful in conjuction with `last_warnings()`, which prints backtraces for all the deprecation warnings issued during the last top-level command.
+First, add a badge to the argument description:
 
-### Test deprecated features
+```{r}
+#' @param na.rm `r lifecycle::badge("deprecated")` `na.rm = FALSE` is no
+#'   longer supported; this function will always remove missing values
+```
 
-Test whether a deprecated feature still works by setting `lifecycle_verbosity` to `"quiet"`:
+And add a deprecation warning if `na.rm` is FALSE.
+In this case, there's no replacement to the behaviour, so we instead use `details` to provide a custom message:
 
-```{r, eval = FALSE}
-test_that("`baz` argument of `foobar_adder()` still works", {
-  withr::local_options(list(lifecycle_verbosity = "quiet"))
-  foobar_adder(1, baz = 2)
-})
+```{r}
+add_two <- function(x, y, na.rm = TRUE) {
+  if (!isTRUE(na.rm)) {
+    lifecycle::deprecate_warn(
+      when = "1.0.0", 
+      what = "add_two(na.rm = )",
+      details = "Ability to remove missing values will be dropped in next release"
+    )
+  }
+  
+  sum(x, y, na.rm = na.rm)
+}
+
+add_two(1, NA, na.rm = TRUE)
+add_two(1, NA, na.rm = FALSE)
 ```
 
-You can also set up verbosity for a whole testthat file within `setup()` and `teardown()` blocks:
+### Deprecating, with new default
 
-```{r, eval = FALSE}
-setup(options(lifecycle_verbosity = "quiet"))
-teardown(options(lifecycle_verbosity = NULL))
+Alternatively, you can change the default value to `lifecycle::deprecated()` to make the deprecation status more obvious from the outside, and use `lifecycle::is_present()`[^1] to test whether or not the argument was provided.
+
+[^1]: Unlike `missing()`, this works for both direct and indirect calls.
+
+```{r}
+#' @importFrom lifecycle deprecated
+add_two <- function(x, y, na.rm = deprecated()) {
+  if (lifecycle::is_present(na.rm)) {
+    lifecycle::deprecate_warn(
+      when = "1.0.0", 
+      what = "add_two(na.rm = )",
+      details = "Ability to remove missing values will be dropped in next release"
+    )
+  }
+  
+  sum(x, y, na.rm = na.rm)
+}
 ```
 
-Test that a feature is correctly deprecated with `expect_deprecated()` or `expect_defunct()`:
+The chief advantage of this technique is that users will get a warning regardless of what value of `na.rm` they use:
```

Comment:
Is the previous example ever desirable? If we want to remove an argument from a function we need to make sure it isn't provided by users. I would consider the last case to be an anti-example.

Reply:
The main problem is that it causes unconditional deprecation warnings, even when the behaviour hasn't changed. That forces package developers to immediately add some `if (packageVersion())` switch, even if the behaviour hasn't actually changed. But maybe that's an argument for going via `deprecate_soft()`?

Reply:
Agreed that `deprecate_soft()` would make sense, then dependencies don't need to send a preemptive release with `if (packageVersion())`. They can bump the dependency version instead of using conditional logic.

It seems that you are describing a kind of "soft" deprecation procedure for removing an argument, with one additional stage where you only deprecate supplied arguments that flip the default. However most users will fall in that category, so that new stage will take a lot of time only for the convenience of the few users who explicitly supplied the default value.